File: VB/WinWebSolution.Module/XpoDataStoreProxy.vb
Imports DevExpress.Xpo.DB
Imports DevExpress.Xpo
Imports DevExpress.Xpo.Metadata
Imports System.Collections.Generic
Imports System.Linq
Imports DevExpress.Xpo.Helpers

Namespace WinWebSolution.Module

    Public Class XpoDataStoreProxy
        Implements IDataStore, ICommandChannel

        Private legacyDataLayer As SimpleDataLayer

        Private legacyDataStore As IDataStore

        Private tempDataLayer As SimpleDataLayer

        Private tempDataStore As IDataStore

        Private tempDatabaseTables As String() = New String() {"ModuleInfo", "XPObjectType"}

        Private Function IsTempDatabaseTable(ByVal tableName As String) As Boolean
            If Not String.IsNullOrEmpty(tableName) Then
                For Each currentTableName As String In tempDatabaseTables
                    If tableName.EndsWith(currentTableName) Then
                        Return True
                    End If
                Next
            End If

            Return False
        End Function

        Public Sub Initialize(ByVal dictionary As XPDictionary, ByVal legacyConnectionString As String, ByVal tempConnectionString As String)
            Dim legacyDictionary As ReflectionDictionary = New ReflectionDictionary()
            Dim tempDictionary As ReflectionDictionary = New ReflectionDictionary()
            For Each ci As XPClassInfo In dictionary.Classes
                If Not IsTempDatabaseTable(ci.TableName) Then
                    legacyDictionary.QueryClassInfo(ci.ClassType)
                Else
                    tempDictionary.QueryClassInfo(ci.ClassType)
                End If
            Next

            legacyDataStore = XpoDefault.GetConnectionProvider(legacyConnectionString, AutoCreateOption.DatabaseAndSchema)
            legacyDataLayer = New SimpleDataLayer(legacyDictionary, legacyDataStore)
            tempDataStore = XpoDefault.GetConnectionProvider(tempConnectionString, AutoCreateOption.DatabaseAndSchema)
            tempDataLayer = New SimpleDataLayer(tempDictionary, tempDataStore)
        End Sub

        Public ReadOnly Property AutoCreateOption As AutoCreateOption Implements IDataStore.AutoCreateOption
            Get
                Return AutoCreateOption.DatabaseAndSchema
            End Get
        End Property

        Public Function ModifyData(ParamArray dmlStatements As ModificationStatement()) As ModificationResult Implements IDataStore.ModifyData
            Dim legacyChanges As List(Of ModificationStatement) = New List(Of ModificationStatement)(dmlStatements.Length)
            Dim tempChanges As List(Of ModificationStatement) = New List(Of ModificationStatement)(dmlStatements.Length)
            For Each stm As ModificationStatement In dmlStatements
                If IsTempDatabaseTable(stm.TableName) Then
                    tempChanges.Add(stm)
                Else
                    legacyChanges.Add(stm)
                End If
            Next

            Dim resultSet As List(Of ParameterValue) = New List(Of ParameterValue)()
            If legacyChanges.Count > 0 Then
                resultSet.AddRange(legacyDataLayer.ModifyData(legacyChanges.ToArray()).Identities)
            End If

            If tempChanges.Count > 0 Then
                resultSet.AddRange(tempDataLayer.ModifyData(tempChanges.ToArray()).Identities)
            End If

            Return New ModificationResult(resultSet)
        End Function

        Public Function SelectData(ParamArray selects As SelectStatement()) As SelectedData Implements IDataStore.SelectData
            Dim isExternals = selects.[Select](Function(stmt) IsTempDatabaseTable(stmt.TableName)).ToList()
            Dim mainSelects As List(Of SelectStatement) = New List(Of SelectStatement)(selects.Length)
            Dim externalSelects As List(Of SelectStatement) = New List(Of SelectStatement)(selects.Length)
            Dim i As Integer = 0
            While i < isExternals.Count
                Call(If(isExternals(i), externalSelects, mainSelects)).Add(selects(i))
                System.Threading.Interlocked.Increment(i)
            End While

            Dim externalResults =(If(externalSelects.Count = 0, Enumerable.Empty(Of SelectStatementResult)(), tempDataLayer.SelectData(externalSelects.ToArray()).ResultSet)).GetEnumerator()
            Dim mainResults =(If(mainSelects.Count = 0, Enumerable.Empty(Of SelectStatementResult)(), legacyDataLayer.SelectData(mainSelects.ToArray()).ResultSet)).GetEnumerator()
            Dim results As SelectStatementResult() = New SelectStatementResult(isExternals.Count - 1) {}
            Dim i As Integer = 0
            While i < results.Length
                Dim enumerator = If(isExternals(i), externalResults, mainResults)
                enumerator.MoveNext()
                results(i) = enumerator.Current
                System.Threading.Interlocked.Increment(i)
            End While

            Return New SelectedData(results)
        End Function

        Public Function UpdateSchema(ByVal dontCreateIfFirstTableNotExist As Boolean, ParamArray tables As DBTable()) As UpdateSchemaResult Implements IDataStore.UpdateSchema
            Dim db1Tables As List(Of DBTable) = New List(Of DBTable)()
            Dim db2Tables As List(Of DBTable) = New List(Of DBTable)()
            For Each table As DBTable In tables
                If Not IsTempDatabaseTable(table.Name) Then
                    db1Tables.Add(table)
                Else
                    db2Tables.Add(table)
                End If
            Next

            legacyDataStore.UpdateSchema(False, db1Tables.ToArray())
            tempDataStore.UpdateSchema(False, db2Tables.ToArray())
            Return UpdateSchemaResult.SchemaExists
        End Function

        Public Function [Do](ByVal command As String, ByVal args As Object) As Object Implements ICommandChannel.[Do]
            Return CType(legacyDataLayer, ICommandChannel).Do(command, args)
        End Function
    End Class
End Namespace

File: XPO/NET.Core/Blazor/AspNetCore.Module/Services/XpoDataStoreProxy.cs
using DevExpress.Xpo;
using DevExpress.Xpo.DB;
using DevExpress.Xpo.Helpers;
using DevExpress.Xpo.Metadata;

namespace AspNetCore.Module.Services;

public class XpoDataStoreProxy : IDataStoreAsync, ICommandChannel {
    private SimpleDataLayer legacyDataLayer;
    private IDataStore legacyDataStore;
    private SimpleDataLayer tempDataLayer;
    private IDataStore tempDataStore;
    private string[] tempDatabaseTables = new string[] { "ModuleInfo", "XPObjectType" };

    private bool IsTempDatabaseTable(string tableName) {
        if(!string.IsNullOrEmpty(tableName)) {
            foreach(string currentTableName in tempDatabaseTables) {
                if(tableName.EndsWith(currentTableName)) {
                    return true;
                }
            }
        }
        return false;
    }
    public void Initialize(XPDictionary dictionary, string legacyConnectionString, string tempConnectionString) {
        ReflectionDictionary legacyDictionary = new ReflectionDictionary();
        ReflectionDictionary tempDictionary = new ReflectionDictionary();
        foreach(XPClassInfo ci in dictionary.Classes) {
            if(!IsTempDatabaseTable(ci.TableName)) {
                legacyDictionary.QueryClassInfo(ci.ClassType);
            }
            else {
                tempDictionary.QueryClassInfo(ci.ClassType);
            }
        }
        legacyDataStore = XpoDefault.GetConnectionProvider(legacyConnectionString, AutoCreateOption.DatabaseAndSchema);
        legacyDataLayer = new SimpleDataLayer(legacyDictionary, legacyDataStore);

        tempDataStore = XpoDefault.GetConnectionProvider(tempConnectionString, AutoCreateOption.DatabaseAndSchema);
        tempDataLayer = new SimpleDataLayer(tempDictionary, tempDataStore);
    }
    public AutoCreateOption AutoCreateOption {
        get {
            return AutoCreateOption.DatabaseAndSchema;
        }
    }
    public ModificationResult ModifyData(params ModificationStatement[] dmlStatements) {
        List<ModificationStatement> legacyChanges = new List<ModificationStatement>(dmlStatements.Length);
        List<ModificationStatement> tempChanges = new List<ModificationStatement>(dmlStatements.Length);
        foreach(ModificationStatement stm in dmlStatements) {
            if(IsTempDatabaseTable(stm.Table.Name)) {
                tempChanges.Add(stm);
            }
            else {
                legacyChanges.Add(stm);
            }
        }

        List<ParameterValue> resultSet = new List<ParameterValue>();
        if(legacyChanges.Count > 0) {
            resultSet.AddRange(legacyDataLayer.ModifyData(legacyChanges.ToArray()).Identities);
        }
        if(tempChanges.Count > 0) {
            resultSet.AddRange(tempDataLayer.ModifyData(tempChanges.ToArray()).Identities);
        }
        return new ModificationResult(resultSet);
    }
    public SelectedData SelectData(params SelectStatement[] selects) {
        var isExternals = selects.Select(stmt => IsTempDatabaseTable(stmt.Table.Name)).ToList();
        List<SelectStatement> mainSelects = new List<SelectStatement>(selects.Length);
        List<SelectStatement> externalSelects = new List<SelectStatement>(selects.Length);
        for(int i = 0; i < isExternals.Count; ++i) {
            (isExternals[i] ? externalSelects : mainSelects).Add(selects[i]);
        }
        var externalResults = (externalSelects.Count == 0 ? Enumerable.Empty<SelectStatementResult>() : tempDataLayer.SelectData(externalSelects.ToArray()).ResultSet).GetEnumerator();
        var mainResults = (mainSelects.Count == 0 ? Enumerable.Empty<SelectStatementResult>() : legacyDataLayer.SelectData(mainSelects.ToArray()).ResultSet).GetEnumerator();
        SelectStatementResult[] results = new SelectStatementResult[isExternals.Count];
        for(int i = 0; i < results.Length; ++i) {
            var enumerator = isExternals[i] ? externalResults : mainResults;
            enumerator.MoveNext();
            results[i] = enumerator.Current;
        }
        return new SelectedData(results);
    }
    public UpdateSchemaResult UpdateSchema(bool dontCreateIfFirstTableNotExist, params DBTable[] tables) {
        List<DBTable> db1Tables = new List<DBTable>();
        List<DBTable> db2Tables = new List<DBTable>();

        foreach(DBTable table in tables) {
            if(!IsTempDatabaseTable(table.Name)) {
                db1Tables.Add(table);
            }
            else {
                db2Tables.Add(table);
            }
        }
        legacyDataStore.UpdateSchema(false, db1Tables.ToArray());
        tempDataStore.UpdateSchema(false, db2Tables.ToArray());
        return UpdateSchemaResult.SchemaExists;
    }
    public object Do(string command, object args) {
        return ((ICommandChannel)legacyDataLayer).Do(command, args);
    }

    public Task<SelectedData> SelectDataAsync(CancellationToken cancellationToken, params SelectStatement[] selects) {
        cancellationToken.ThrowIfCancellationRequested();
        return Task.FromResult(SelectData(selects));
    }

    public Task<ModificationResult> ModifyDataAsync(CancellationToken cancellationToken, params ModificationStatement[] dmlStatements) {
        cancellationToken.ThrowIfCancellationRequested();
        return Task.FromResult(ModifyData(dmlStatements));
    }

    public Task<UpdateSchemaResult> UpdateSchemaAsync(CancellationToken cancellationToken, bool doNotCreateIfFirstTableNotExist, params DBTable[] tables) {
        cancellationToken.ThrowIfCancellationRequested();
        return Task.FromResult(UpdateSchema(doNotCreateIfFirstTableNotExist, tables));
    }
}
